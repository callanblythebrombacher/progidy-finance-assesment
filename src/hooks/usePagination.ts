import {useState} from 'react';

/**
 * Custom hook to manage pagination logic.
 * @param tableRowsLength The total number of rows in the table.
 * @param page The current page index.
 * @returns An object containing pagination state and functions.
 */
const usePagination = (tableRowsLength: number, page: number) => {
  const [numberOfItemsPerPageList] = useState([15, 30, 60]); // Array of items per page options
  const [itemsPerPage, setItemsPerPage] = useState(numberOfItemsPerPageList[0]); // State for items per page

  // Calculate the starting index (from) and ending index (to) of the displayed rows on the current page
  const from = page * itemsPerPage;
  const to = Math.min((page + 1) * itemsPerPage, tableRowsLength);

  return {
    numberOfItemsPerPageList, // List of available items per page options
    itemsPerPage, // Current number of items per page
    setItemsPerPage, // Function to set the number of items per page
    from, // Starting index of displayed rows on the current page
    to, // Ending index of displayed rows on the current page
  };
};

export default usePagination;
