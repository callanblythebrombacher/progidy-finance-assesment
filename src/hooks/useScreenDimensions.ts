import {useEffect, useState} from 'react';
import {Dimensions} from 'react-native';

interface ScreenDimensions {
  height: number;
  width: number;
}

const useScreenDimensions = (): ScreenDimensions => {
  const [screenDimensions, setScreenDimensions] = useState<ScreenDimensions>({
    height: Dimensions.get('window').height,
    width: Dimensions.get('window').width,
  });

  useEffect(() => {
    const updateScreenDimensions = ({window: {height, width}}: any) => {
      setScreenDimensions({height, width});
    };

    Dimensions.addEventListener('change', updateScreenDimensions);
  }, []);

  return screenDimensions;
};

export default useScreenDimensions;
