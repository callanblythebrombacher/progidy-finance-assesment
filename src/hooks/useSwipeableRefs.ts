import {useRef} from 'react';
import {SwipeableComponent} from '../components/atoms/swipeable.atom.tsx';

const useSwipeableRefs = (length: number) => {
  const refsArray = useRef<
    (React.ElementRef<typeof SwipeableComponent> | null)[]
  >([]);

  return refsArray;
};

export default useSwipeableRefs;
