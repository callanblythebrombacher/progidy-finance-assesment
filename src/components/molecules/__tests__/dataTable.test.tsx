import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {Table} from '../dataTable.molecule';
import {resetDeletionEvent} from '../../../redux/reducers/country.slice';
import usePagination from '../../../hooks/usePagination';
import useSwipeableRefs from '../../../hooks/useSwipeableRefs';
import useScreenDimensions from '../../../hooks/useScreenDimensions';
import {useSelector, useDispatch} from 'react-redux';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

jest.mock('../../../hooks/usePagination');
jest.mock('../../../hooks/useSwipeableRefs');
jest.mock('../../../hooks/useScreenDimensions');
jest.mock('../../atoms/tableHeader.atom', () => ({
  TableHeader: () => 'TableHeader',
}));
jest.mock('../../atoms/tableRow.atom', () => ({
  TableRow: () => 'TableRow',
}));
jest.mock('../../atoms/swipeable.atom', () => ({
  SwipeableComponent: jest.fn(({children}) => children),
}));

describe('Table Component', () => {
  const mockDispatch = jest.fn();
  const mockUseSelector = useSelector as jest.Mock;
  const mockUseDispatch = useDispatch as jest.Mock;

  const mockHeaderRow = [{item: 'Header1', isNumeric: false}];
  const mockTableRows = [
    {data: [{item: 'Row1', isNumeric: false}], isActive: true},
    {data: [{item: 'Row2', isNumeric: false}], isActive: true},
  ];
  const mockHandleDelete = jest.fn();
  const mockSetPage = jest.fn();

  beforeEach(() => {
    mockUseDispatch.mockReturnValue(mockDispatch);
    mockUseSelector.mockReturnValue(false);
    (usePagination as jest.Mock).mockReturnValue({
      from: 0,
      to: 1,
      numberOfItemsPerPageList: [10, 20, 30],
      itemsPerPage: 10,
      setItemsPerPage: jest.fn(),
    });
    (useSwipeableRefs as jest.Mock).mockReturnValue({current: []});
    (useScreenDimensions as jest.Mock).mockReturnValue({height: 800});
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with given props', () => {
    const {toJSON} = render(
      <Table
        headerRow={mockHeaderRow}
        tableRows={mockTableRows}
        handleDelete={mockHandleDelete}
        page={0}
        setPage={mockSetPage}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders TableHeader and TableRow components', () => {
    const {getByText} = render(
      <Table
        headerRow={mockHeaderRow}
        tableRows={mockTableRows}
        handleDelete={mockHandleDelete}
        page={0}
        setPage={mockSetPage}
      />,
    );

    expect(getByText('TableHeader')).toBeTruthy();
    expect(getByText('TableRow')).toBeTruthy(); // Assumes TableRow renders some identifiable content
  });

  it('renders SwipeableComponent for each row', () => {
    const {getAllByText} = render(
      <Table
        headerRow={mockHeaderRow}
        tableRows={mockTableRows}
        handleDelete={mockHandleDelete}
        page={0}
        setPage={mockSetPage}
      />,
    );

    expect(getAllByText('SwipeableComponent')).toHaveLength(
      mockTableRows.length,
    );
  });

  it('calls handleDelete when delete action is triggered', () => {
    const {getByText} = render(
      <Table
        headerRow={mockHeaderRow}
        tableRows={mockTableRows}
        handleDelete={mockHandleDelete}
        page={0}
        setPage={mockSetPage}
      />,
    );

    fireEvent.press(getByText('Row1'));
    expect(mockHandleDelete).toHaveBeenCalledWith('Row1');
  });

  it('resets page and dispatches resetDeletionEvent on tableRows change', () => {
    render(
      <Table
        headerRow={mockHeaderRow}
        tableRows={mockTableRows}
        handleDelete={mockHandleDelete}
        page={0}
        setPage={mockSetPage}
      />,
    );

    expect(mockSetPage).toHaveBeenCalledWith(0);
    expect(mockDispatch).toHaveBeenCalledWith(resetDeletionEvent());
  });

  it('renders DataTable.Pagination with correct props', () => {
    const {getByTestId} = render(
      <Table
        headerRow={mockHeaderRow}
        tableRows={mockTableRows}
        handleDelete={mockHandleDelete}
        page={0}
        setPage={mockSetPage}
      />,
    );

    const pagination = getByTestId('data-table-pagination');
    expect(pagination.props.page).toBe(0);
    expect(pagination.props.numberOfPages).toBe(1);
    expect(pagination.props.numberOfItemsPerPageList).toEqual([10, 20, 30]);
    expect(pagination.props.numberOfItemsPerPage).toBe(10);
  });
});
