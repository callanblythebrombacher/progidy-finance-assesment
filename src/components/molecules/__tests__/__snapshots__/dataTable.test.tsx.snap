// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component renders correctly with given props 1`] = `
<View
  style={
    [
      {
        "width": "100%",
      },
      {
        "flex": 1,
        "position": "relative",
        "zIndex": 1,
      },
    ]
  }
>
  TableHeader
  <RCTScrollView
    contentInsetAdjustmentBehavior="automatic"
    style={
      [
        {
          "backgroundColor": "white",
        },
        {
          "height": 576,
        },
      ]
    }
  >
    <View>
      TableRow
    </View>
  </RCTScrollView>
</View>
`;
